Modules/_testbuffer.c

```diff
@@ -2813,48 +2813,27 @@ static struct PyMethodDef _testbuffer_functions[] = {
     {NULL, NULL}
 };
 
-static struct PyModuleDef _testbuffermodule = {
-    PyModuleDef_HEAD_INIT,
-    "_testbuffer",
-    NULL,
-    -1,
-    _testbuffer_functions,
-    NULL,
-    NULL,
-    NULL,
-    NULL
-};
-
-
-PyMODINIT_FUNC
-PyInit__testbuffer(void)
+static int
+_testbuffer_exec(PyObject *m)
 {
-    PyObject *m;
-
-    m = PyModule_Create(&_testbuffermodule);
-    if (m == NULL)
-        return NULL;
-
-    Py_SET_TYPE(&NDArray_Type, &PyType_Type);
-    Py_INCREF(&NDArray_Type);
-    PyModule_AddObject(m, "ndarray", (PyObject *)&NDArray_Type);
+    if (PyModule_AddType(m, &NDArray_Type) < 0)
+        return -1;
 
-    Py_SET_TYPE(&StaticArray_Type, &PyType_Type);
-    Py_INCREF(&StaticArray_Type);
-    PyModule_AddObject(m, "staticarray", (PyObject *)&StaticArray_Type);
+    if (PyModule_AddType(m, &StaticArray_Type) < 0)
+        return -1;
 
     structmodule = PyImport_ImportModule("struct");
     if (structmodule == NULL)
-        return NULL;
+        return -1;
 
     Struct = PyObject_GetAttrString(structmodule, "Struct");
     calcsize = PyObject_GetAttrString(structmodule, "calcsize");
     if (Struct == NULL || calcsize == NULL)
-        return NULL;
+        return -1;
 
     simple_format = PyUnicode_FromString(simple_fmt);
```

Comment:
Hum, would it be possible to move these static variables to be able to clear them when the extension module is unloaded? It would require to pass the module to NDArray_Type creation, so be able to get the module state from a ndarray instance.

```
static const char *simple_fmt = "B";
static PyObject *simple_format = NULL;
```

Reply:
@vstinner This seems tricky.  How do you recommend passing the module to a static type like NDArray_Type.  I can't put the module instance on the type or it will be overridden each time we do _testbuffer_exec.  Is there a way to do this in tp_new?

Reply:
@vstinner What if I just call PyUnicode_FromString every time it is used, since as you say this is just a test module.  Struct and calcsize should be handled properly in any case.

Reply:
Usually to convert an extension module to multiphase init, you must convert static types to heap types first.

Reply:
@vstinner see my comment in response to @shihai1991 about this - I was not able to convert these to heap types